src/future/future/flatten.rs

```diff
@@ -0,0 +1,46 @@
+use futures_core::ready;
+use std::pin::Pin;
+
+use crate::future::Future;
+use crate::future::IntoFuture;
+use crate::task::{Context, Poll};
+
+#[doc(hidden)]
+#[derive(Debug)]
+pub enum FlattenFuture<Fut1, Fut2> {
```

Comment:
Also, if using enum, don't want the state to be public, so I think it is nice to be wrapped in a struct.

```rust
pub struct FlattenFuture<Fut1, Fut2> {
    state: State<Fut1, Fut2>,
}

enum State<Fut1, Fut2> {
    First(Fut1),
    Second(Fut2),
    Empty,
}
```